actors/builtin/power/power_actor.go

```diff
@@ -590,10 +414,79 @@ func (a Actor) OnEpochTickEnd(rt Runtime, _ *adt.EmptyValue) *adt.EmptyValue {
 	return nil
 }
 
+func (a Actor) updatePledgeCollateralTally(rt Runtime, pledgeAmount abi.TokenAmount) error {
+	var st State
+	rt.State().Transaction(&st, func() interface{} {
+		st.TotalPledgeCollateral = big.Add(st.TotalPledgeCollateral, pledgeAmount)
+		Assert(st.TotalPledgeCollateral.GreaterThanEqual(big.Zero()))
+
+		return nil
+	})
+	return nil
+}
+
+func (a Actor) UpdatePledgeCollateralTally(rt Runtime, pledgeAmount abi.TokenAmount) *adt.EmptyValue {
+	rt.ValidateImmediateCallerType(builtin.StorageMinerActorCodeID)
+	a.updatePledgeCollateralTally(rt, pledgeAmount)
+	return nil
+}
+
+func (a Actor) OnConsensusFault(rt Runtime, pledgeAmount abi.TokenAmount) *adt.EmptyValue {
+	rt.ValidateImmediateCallerType(builtin.StorageMinerActorCodeID)
+
+	minerAddr := rt.Message().Caller()
+	nominal, ok := rt.ResolveAddress(minerAddr)
+	Assert(ok)
+
+	var st State
+	rt.State().Transaction(&st, func() interface{} {
+
+		claim, powerOk, err := st.getClaim(adt.AsStore(rt), nominal)
+		if err != nil {
+			rt.Abortf(exitcode.ErrIllegalState, "failed to read claimed power for fault: %v", err)
+		}
+		if !powerOk {
+			rt.Abortf(exitcode.ErrIllegalArgument, "miner %v not registered (already slashed?)", nominal)
+		}
+		Assert(claim.RawBytePower.GreaterThanEqual(big.Zero()))
+		Assert(claim.QualityAdjPower.GreaterThanEqual(big.Zero()))
+
+		st.TotalQualityAdjPower = big.Sub(st.TotalQualityAdjPower, claim.QualityAdjPower)
+		st.TotalRawBytePower = big.Sub(st.TotalRawBytePower, claim.RawBytePower)
+
+		return nil
+	})
+
+	err := a.updatePledgeCollateralTally(rt, pledgeAmount)
```

Comment:
This should be subtracting the former miner's pledge, not adding it.

Since you already have a transaction right above here, delete `updatePledgeCollateralTally` (inline it in `UpdatePledgeCollateralTally `) and do the update in the existing transaction.

Reply:
pledgeAmount has been negated before invoking this method, moving the negation into this file and adding change as suggested.